src/activity_handlers.cpp

```diff
@@ -1501,6 +1502,51 @@ void activity_handlers::repair_item_finish( player_activity *act, player *p )
     act->moves_left = actor->move_cost;
 }
 
+void activity_handlers::mend_item_finish( player_activity *act, player *p )
+{
+    item_location target;
+    switch( static_cast<item_location::type>( act->index ) ) {
```

Comment:
Your current solution exposes implementation details (the `type`, position) and makes modifying/extending the class more complicated. `item_location` itself should be responsible for that task.

We need some way to store an `item_location` in a `player_activity`. Ideally we should be able to store several of them (for example to repair item X using item Y), a `vector<item_location>` seems appropriate.

I suggest this: change `item_location` to deriving from `JsonSerializer` and `JsonDeserializer`, see other class like npc/player/item/...). Store item locations in a vector in the activity and serialized the vector the same way the other two vectors are serialized.

Serializing is relatively simple (see savegame_json.cpp), as is deserialization. You may need to add error checking and special handling for the item pointer. The pointer itself can not be stored, you need to store some abstract reference (e.g. the index in the item list) and recalculated the pointer after deserialization. Deserailization may result in a null-location, but this needs to be handled by the activity anyway as the item may get lost during the activity.

This allows extending / modifying `item_location` (even adding new types, like in the inventory of pets) without any changes outside of that class / that cpp file. It's also usable in other parts of the game and allows storing several locations in the `player_activity`.

In short: a user of the `item_location` shouldn't need to know about the internals, they only need to know about the serialize/deserialize functions.